AsyncDisplayKit/Private/_ASHierarchyChangeSet.m

```diff
@@ -277,6 +301,83 @@ - (void)_sortAndCoalesceChangeArrays
+  
+  // Assert that no invalid deletes/reloads happened.
+  NSInteger invalidSectionDelete = NSNotFound;
+  if (oldSectionCount == 0) {
+    invalidSectionDelete = [_deletedSections firstIndex];
+  } else {
+    invalidSectionDelete = [_deletedSections indexGreaterThanIndex:oldSectionCount - 1];
+  }
+  ASDisplayNodeAssert(NSNotFound == invalidSectionDelete, @"Attempt to delete section %ld but there are only %ld sections before the update.", invalidSectionDelete, oldSectionCount);
+  
+  for (_ASHierarchyItemChange *change in _deleteItemChanges) {
+    for (NSIndexPath *indexPath in change.indexPaths) {
+      // Assert that item delete happened in a valid section.
+      ASDisplayNodeAssert(indexPath.section < oldSectionCount, @"Attempt to delete item %ld from section %ld, but there are only %ld sections before the update.", (long)indexPath.item, (long)indexPath.section, (long)oldSectionCount);
+      
+      // Assert that item delete happened to a valid item.
+      NSInteger oldItemCount = _oldItemCounts[indexPath.section].integerValue;
+      ASDisplayNodeAssert(indexPath.item < oldItemCount, @"Attempt to delete item %ld from section %ld, which only contains %ld items before the update.", (long)indexPath.item, (long)indexPath.section, (long)oldItemCount);
+    }
+  }
+  
+  for (_ASHierarchyItemChange *change in _insertItemChanges) {
+    for (NSIndexPath *indexPath in change.indexPaths) {
+      // Assert that item insert happened in a valid section.
+      ASDisplayNodeAssert(indexPath.section < newSectionCount, @"Attempt to insert item %ld into section %ld, but there are only %ld sections after the update.", (long)indexPath.item, (long)indexPath.section, (long)newSectionCount);
+      
+      // Assert that item delete happened to a valid item.
+      NSInteger newItemCount = _newItemCounts[indexPath.section].integerValue;
+      ASDisplayNodeAssert(indexPath.item < newItemCount, @"Attempt to insert item %ld into section %ld, which only contains %ld items after the update.", (long)indexPath.item, (long)indexPath.section, (long)newItemCount);
+    }
+  }
+  
+  // Assert that no sections were inserted out of bounds.
+  NSInteger invalidSectionInsert = NSNotFound;
+  if (newSectionCount == 0) {
+    invalidSectionInsert = [_insertedSections firstIndex];
+  } else {
+    invalidSectionInsert = [_insertedSections indexGreaterThanIndex:newSectionCount - 1];
+  }
+  ASDisplayNodeAssert(NSNotFound == invalidSectionInsert, @"Attempt to insert section %ld but there are only %ld sections after the update.", (long)invalidSectionInsert, (long)newSectionCount);
+  
+  [_oldItemCounts enumerateObjectsUsingBlock:^(NSNumber * _Nonnull oldItemCountObj, NSUInteger oldSection, BOOL * _Nonnull stop) {
```

Comment:
@appleguy I thought in this case that the readability benefit was worth the performance cost but I'll change it.